optaplanner-docs/src/modules/ROOT/pages/constraint-streams/constraint-streams.adoc

```diff
@@ -963,6 +966,51 @@ from the given facts.
 Using this method, you ensure that the constraint provider does not miss any constraints and that the scoring function
 remains consistent as your code base evolves.
 
+[[constraintStreamsTestingWhatActuallyHappens]]
+==== What actually happens
+
+The purpose of `ConstraintVerifier` is to test constraints in an environment as close as possible to the production deployment.
+Therefore when the `penalizesBy(...)` method (or similar) is called, the following will happen:
+
+* The constraint(s) provided by `verifyThat(...)` method are compiled and the new solver session is created,
+much like if the `Solver` or `SolverManager` APIs were used.
+* All the facts provided by the `given(...)` method are inserted into the solver session.
+* A score will be calculated, executing the constraint(s) on the given facts.
+This will include execution of all variable listeners, should the planning entities have any.
+
+For example, consider the "distanceToPreviousStandstill" constraint in the xref:use-cases-and-examples/vehicle-routing/vehicle-routing.adoc[vehicle routing example]:
+
+[source,java,options="nowrap"]
+----
+    Constraint distanceToPreviousStandstill(ConstraintFactory factory) {
+        return factory.forEach(Customer.class)
+                .penalizeLong("distanceToPreviousStandstill",
+                        HardSoftLongScore.ONE_SOFT,
+                        Customer::getDistanceFromPreviousStandstill);
+    }
+----
+
+Notice how the constraint does not make any `Vehicle` references.
+Yet the `ConstraintVerifier`-based test provides `Vehicle` instances:
+
+[source,java,options="nowrap"]
+----
+    @Test
+    void distanceToPreviousStandstill() {
+        Vehicle vehicleA = ...
+        Customer customer1 = ...
+        Customer customer2 = ...
+
+        constraintVerifier.verifyThat(VehicleRoutingConstraintProvider::distanceToPreviousStandstill)
+                .given(vehicleA, customer1, customer2)
+                .penalizesBy(9000L);
+    }
+----
+
+This is necessary, as the `Customer` class has a shadow variable based on `Vehicle`.
+Had no `Vehicle` instances been provided through the `given(...)` call,
+the constraint would not match anything and the test would have failed.
```

Comment:
This whole section was triggered by this thread:
https://kie.zulipchat.com/#narrow/stream/232679-optaplanner/topic/Testing.20ConstraintProviders/near/272108899

But as you are the expert on variable listeners by now, perhaps you have a better way of explaining, @yurloc ? I'm not too happy with this wording.

Reply:
Incorrect. Vehicle doesn't have to be inserted here. The test still passes with the following modification:
```diff
diff --git a/optaplanner-examples/src/test/java/org/optaplanner/examples/vehiclerouting/score/VehicleRoutingConstraintProviderTest.java b/optaplanner-examples/src/test/java/org/optaplanner/examples/vehiclerouting/score/VehicleRoutingConstraintProviderTest.java
index 8202d01ee3..883c2eb5f6 100644
--- a/optaplanner-examples/src/test/java/org/optaplanner/examples/vehiclerouting/score/VehicleRoutingConstraintProviderTest.java
+++ b/optaplanner-examples/src/test/java/org/optaplanner/examples/vehiclerouting/score/VehicleRoutingConstraintProviderTest.java
@@ -73,15 +73,11 @@ class VehicleRoutingConstraintProviderTest {
         Vehicle vehicleA = new Vehicle(1L, 100, new Depot(1L, location1));
         Customer customer1 = new Customer(2L, location2, 80);
         customer1.setPreviousStandstill(vehicleA);
-        customer1.setVehicle(vehicleA);
-        vehicleA.setNextCustomer(customer1);
         Customer customer2 = new Customer(3L, location3, 40);
         customer2.setPreviousStandstill(customer1);
-        customer2.setVehicle(vehicleA);
-        customer1.setNextCustomer(customer2);
 
         constraintVerifier.verifyThat(VehicleRoutingConstraintProvider::distanceToPreviousStandstill)
-                .given(vehicleA, customer1, customer2)
+                .given(customer1, customer2)
                 .penalizesBy(9000L);
     }
 ```
Plus, no listeners are being triggered, so if you don't initialize shadow variables manually before executing `verifyThat(...).given(...).penalizesBy(...)`, they will remain uninitialized.

I think the user on Zulip didn't properly construct the tested problem.

Reply:
If the vehicle wasn't reachable via references from the _given()_ customers, then it was definitely not in the working memory and _giving()_ it explicitly could make a difference. Depends on what the constraint looks like.

Maybe the user both connected the vehicle to customers and _gave()_ it to the verifier and then misinterpreted what actually fixed it.

Anyway, listeners are not involved when using `given(Object... facts)` because a score director is only build when using `given(Solution_ solution)`:

https://github.com/kiegroup/optaplanner/blob/8ef196252291d00cdf6a033ea3ade9927a919fbf/optaplanner-test/src/main/java/org/optaplanner/test/impl/score/stream/DefaultSingleConstraintVerification.java#L38-L50

Reply:
I'm not saying the documentation shouldn't be improved. I just think the listeners thing is a red herring and the currently proposed wording is not accurate.